### DISCUSSION TITLE
What security escape issues will non-root container users encounter when using --privileged or SYS_ADMIN (docker run --privileged --user non-root)

### DISCUSSION THREAD

**Discussion by yushiqie** [moby/moby#48667] [Q&A] [answered]
If a container is run with privileged mode or with --cap-add=SYS_ADMIN using docker run , but the user inside the container is non-root, what kind of security escape risks will be encountered, particularly to the host, and are there any ways to mitigate such risks

**Comment by yushiqie** [comment]
My purpose is to run containerd-rootless/runc-rootless in a Docker container , must with the requirement of either using docker run --privileged or docker run --cap-add=SYS_ADMIN. Presently, Docker also facilitates rootless Docker-in-Docker via https://docs.docker.com/engine/security/rootless/#rootless-docker-in-docker; however, this also necessitates operating in privileged mode. I'm curious if this approach introduces any escape risk, and if there are any best practices you would recommend to mitigate such concerns. Your insights are greatly appreciated

**Comment by AkihiroSuda** [comment, marked as answer]
Basically it is just same as a non-root user on the host.
What you can't do with a non-root user on the host isn't possible with `docker run --privileged --user <NONROOT>` either.

**Reply by yushiqie** [reply]
What is the corresponding non-root user on the host for a non-root user inside a container? How can I check what permissions the non-root user has on the host? Can --privileged potentially lead to privilege escalation, where the non-root user on the host gains root user permissions, thereby causing some security risks?

**Comment by nggit** [comment]
To mitigate the risks of `--cap-add=SYS_ADMIN`, after you've done something that requires it, like `mount`, `iptables`, or whatever,
you can use [capsh](https://man7.org/linux/man-pages/man1/capsh.1.html) to invoke the main program, while dropping the dangerous capabilities.

```
exec capsh --drop=cap_sys_admin,cap_net_admin,cap_net_raw,cap_setfcap,cap_setpcap -- -c '/your/main/prog'
```

> but the user inside the container is non-root, 

This is good. You can experiment with the `--user` flag in `capsh`.

**Reply by yushiqie** [reply]
"What I want to do is run containerd-rootless inside a Docker container, which requires Docker --privileged permissions. The container application running inside containerd can also set Capabilities via the OCI config to achieve the functionality of capsh, thereby restricting the capabilities of the container application. However, for the host Docker, if a security escape occurs with docker run --user nonroot --privileged, what is the greatest risk to the host? Is it equivalent to what a nonroot user on the host can do?"

**Reply by nggit** [reply]
**Yes** with `--privileged`, it will impose the same risk, the damage inflicted by a non-root user on a container is on the same level as a non-root on a Docker host.. **only** if they find a way of root escalation, either with `sudo`, or based on CVE.

**Reply by yushiqie** [reply]
yeap，I got it，thanks😁
